### DISCUSSION TITLE
Remove params arguments option

### DISCUSSION THREAD

**Discussion by wdnijdam** [dotnet/csharplang#7843] [Language Ideas]
I propose to remove the params arguments option from the c# language.

The c# language allows using the params keyword to enlist a set as tail arguments to an array:

`public static void UseParams(params int[] list)`

The problem with this feature is that it may lead to unclarities and programming errors. With the new array initializers, it would be fairly simple to write the list of arguments without using the params feature

`UseParams(["a", 1])` instead of `UseParams("a", 1)`

The params syntax is dangerous ambigue in cases like with EntitityFramework `db.Items.FindAsync(id, cancellationToken)` (wrong) instead of `db.Items.FindAsync([id], cancellationToken)` (right)

**Comment by sharpchen** [comment]
It's a breaking change, even if it's the best practice what we can do is just add a warning diagnostic

**Reply by wdnijdam** [reply]
Yes opting out or have a code analyser warn would help. It is indeed breaking existing code. 

Would it be possible to auto-convert the source code to the collection insertion pattern?

**Reply by yaakov-h** [reply]
Code Analyzers can include Code Fixes which can be used to modify the original source to comply with the analyzer rule. So, yes.

**Comment by CyrusNajmabadi** [comment]
This would break the world badly. If you would like to disallow `params` in your own codebases though, you can do so by writing an analyzer to block it.
